kafkareceiver and kafkaexporter encoding config does not support named extensions

### Component(s)

exporter/kafka, receiver/kafka

### What happened?

## Description

The Kafka exporter and receiver support `encoding` configuration, which may identify an encoding extension. This currently only supports unnamed encoding extensions, e.g. "otlp_encoding", but not "otlp_encoding/json".

## Steps to Reproduce

Run the collector with the attached config.

## Expected Result

It should run.

## Actual Result

The collector fails to start.

### Collector version

7396e408a0b81cc0bf86b09ca1a023fb6c9dadeb

### Environment information

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")


### OpenTelemetry Collector configuration

```yaml
extensions:
  otlp_encoding/json:
    protocol: otlp_json

receivers:
  kafka:
    brokers: [localhost:9093]
    encoding: otlp_encoding/json
  kafkametrics:
    brokers: [localhost:9093]
    collection_interval: 10s
    scrapers: [topics, consumers]

exporters:
  debug:
    verbosity: detailed
  kafka:     
    brokers: [localhost:9093]
    encoding: otlp_encoding/json

service:
  extensions: [otlp_encoding/json]
  pipelines:
    metrics/kafka:
      receivers: [kafka]
      exporters: [debug]
    metrics:
      receivers: [kafkametrics]
      exporters: [kafka]
```

### Log output

```shell
cd ./cmd/otelcontribcol && GO111MODULE=on go run --race . --config ../../local/config.yaml 
2025-05-20T11:51:19.234+0800    info    service@v0.126.1-0.20250515040533-97a6accbc082/service.go:199   Setting up own telemetry...     {"resource": {}}
2025-05-20T11:51:19.235+0800    info    builders/builders.go:26 Development component. May change in the future.        {"resource": {}, "otelcol.component.id": "debug", "otelcol.component.kind": "exporter", "otelcol.signal": "metrics"}
2025-05-20T11:51:19.238+0800    info    service@v0.126.1-0.20250515040533-97a6accbc082/service.go:266   Starting otelcontribcol...      {"resource": {}, "Version": "0.126.0-dev", "NumCPU": 14}
2025-05-20T11:51:19.238+0800    info    extensions/extensions.go:41     Starting extensions...  {"resource": {}}
2025-05-20T11:51:19.238+0800    info    extensions/extensions.go:45     Extension is starting...        {"resource": {}, "otelcol.component.id": "otlp_encoding/json", "otelcol.component.kind": "extension"}
2025-05-20T11:51:19.238+0800    info    extensions/extensions.go:62     Extension started.      {"resource": {}, "otelcol.component.id": "otlp_encoding/json", "otelcol.component.kind": "extension"}
2025-05-20T11:51:19.238+0800    info    extensions/extensions.go:45     Extension is starting...        {"resource": {}, "otelcol.component.id": "otlp_encoding", "otelcol.component.kind": "extension"}
2025-05-20T11:51:19.238+0800    info    extensions/extensions.go:62     Extension started.      {"resource": {}, "otelcol.component.id": "otlp_encoding", "otelcol.component.kind": "extension"}
2025-05-20T11:51:19.240+0800    error   graph/graph.go:438      Failed to start component       {"resource": {}, "error": "unrecognized metrics encoding \"otlp_encoding/json\"", "type": "Receiver", "id": "kafka"}
2025-05-20T11:51:19.240+0800    info    service@v0.126.1-0.20250515040533-97a6accbc082/service.go:331   Starting shutdown...    {"resource": {}}
2025-05-20T11:51:19.241+0800    info    extensions/extensions.go:69     Stopping extensions...  {"resource": {}}
2025-05-20T11:51:19.241+0800    info    service@v0.126.1-0.20250515040533-97a6accbc082/service.go:345   Shutdown complete.      {"resource": {}}
Error: cannot start pipelines: failed to start "kafka" receiver: unrecognized metrics encoding "otlp_encoding/json"
2025/05/20 11:51:19 collector server run finished with error: cannot start pipelines: failed to start "kafka" receiver: unrecognized metrics encoding "otlp_encoding/json"
exit status 1
make: *** [Makefile:333: run] Error 1
```

### Additional context

Related issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39809

--- a/.chloggen/kafka-encoding-name.yaml
+++ b/.chloggen/kafka-encoding-name.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: kafkareceiver, kafkaexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add support for named encoding extensions in kafkareceiver and kafkaexporter
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [40142]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user]

--- a/exporter/kafkaexporter/marshaler.go
+++ b/exporter/kafkaexporter/marshaler.go
@@ -97,12 +97,11 @@ func loadEncodingExtension[T any](host component.Host, encoding, signalType stri
 	return marshaler, nil
 }
 
-// encodingToComponentID converts an encoding string to a component ID using the given encoding as type.
+// encodingToComponentID attempts to parse the encoding string as a component ID.
 func encodingToComponentID(encoding string) (*component.ID, error) {
-	componentType, err := component.NewType(encoding)
-	if err != nil {
-		return nil, fmt.Errorf("invalid component type: %w", err)
+	var id component.ID
+	if err := id.UnmarshalText([]byte(encoding)); err != nil {
+		return nil, fmt.Errorf("invalid component ID: %w", err)
 	}
-	id := component.NewID(componentType)
 	return &id, nil
 }

--- a/exporter/kafkaexporter/marshaler_test.go
+++ b/exporter/kafkaexporter/marshaler_test.go
@@ -34,6 +34,17 @@ func TestGetLogsMarshaler(t *testing.T) {
 	require.Len(t, messages, 1)
 	assert.Equal(t, "overridden", string(messages[0].Value))
 
+	// Verify named extensions are supported.
+	m = mustGetLogsMarshaler(t, "otlp_proto/alice", extensionsHost{
+		component.MustNewIDWithName("otlp_proto", "alice"): plogMarshalerFuncExtension(func(plog.Logs) ([]byte, error) {
+			return []byte("bob"), nil
+		}),
+	})
+	messages, err = m.MarshalLogs(plog.NewLogs())
+	require.NoError(t, err)
+	require.Len(t, messages, 1)
+	assert.Equal(t, "bob", string(messages[0].Value))
+
 	// Specifying an extension for a different type should fail fast.
 	m, err = getLogsMarshaler("otlp_proto", extensionsHost{
 		component.MustNewID("otlp_proto"): struct{ component.Component }{},

--- a/receiver/kafkareceiver/encoding.go
+++ b/receiver/kafkareceiver/encoding.go
@@ -127,12 +127,11 @@ func loadEncodingExtension[T any](host component.Host, encoding, signalType stri
 	return unmarshaler, nil
 }
 
-// encodingToComponentID converts an encoding string to a component ID using the given encoding as type.
+// encodingToComponentID attempts to parse the encoding string as a component ID.
 func encodingToComponentID(encoding string) (*component.ID, error) {
-	componentType, err := component.NewType(encoding)
-	if err != nil {
+	var id component.ID
+	if err := id.UnmarshalText([]byte(encoding)); err != nil {
 		return nil, fmt.Errorf("%w: %w", errInvalidComponentType, err)
 	}
-	id := component.NewID(componentType)
 	return &id, nil
 }

--- a/receiver/kafkareceiver/encoding_test.go
+++ b/receiver/kafkareceiver/encoding_test.go
@@ -163,6 +163,12 @@ func TestNewLogsUnmarshalerExtension(t *testing.T) {
 	})
 	assert.Equal(t, &customLogsUnmarshalerExtension, u)
 
+	// Verify extensions can be named.
+	u = mustNewLogsUnmarshaler(t, "otlp_proto/alice", extensionsHost{
+		component.MustNewIDWithName("otlp_proto", "alice"): &customLogsUnmarshalerExtension,
+	})
+	assert.Equal(t, &customLogsUnmarshalerExtension, u)
+
 	// Specifying an extension for a different type should fail fast.
 	u, err := newLogsUnmarshaler("not_logs", settings, extensionsHost{
 		component.MustNewID("not_logs"): &customTracesUnmarshalerExtension,

